go-controller/pkg/ovn/dns-name-resolver/egressfirewall_external_dns.go

```diff
@@ -0,0 +1,386 @@
+package dnsnameresolver
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"reflect"
+	"sync"
+	"time"
+
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+
+	ocpnetworkapiv1alpha1 "github.com/openshift/api/network/v1alpha1"
+	ocpnetworklisterv1alpha1 "github.com/openshift/client-go/network/listers/network/v1alpha1"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/config"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/controller"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/factory"
+	addressset "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/ovn/address_set"
+)
+
+// ExternalEgressDNS keeps track of DNS names and the corresponding IP addresses.
+// For each DNS name, an address set is allocated and the address set is
+// kept updated with the corresponding IP addresses. Whenever a DNS name
+// is removed the corresponding addresset is destroyed.
+type ExternalEgressDNS struct {
+	// dnsLock protects the operations performed on the maps and address sets.
+	dnsLock sync.Mutex
+	// dnsNames maps the dnsNames to the dnsResolvedNames.
+	dnsNames map[string]dnsResolvedName
+	// namespaceDNSNames maps the DNS names used in the EgressFirewall rule of
+	// a namespace to the namespace
+	namespaceDNSNames map[string]sets.Set[string]
+	// addressSetFactory helps in the creation of the address sets.
+	addressSetFactory addressset.AddressSetFactory
+	// controllerName gives the name of the controller which created this
+	// instance of the ExternalEgressDNS.
+	controllerName string
+
+	// Controller used to handle dns name resolver resoures
+	dnsNameResolverController controller.Controller
+	// Lister for dns name resolver
+	dnsLister ocpnetworklisterv1alpha1.DNSNameResolverLister
+	// maps to hold the details of the DNS names and the corresponding
+	// DNSNameResolver object names
+	dnsNameToResolver map[string]string
+	resolverToDNSName map[string]string
+}
+
+var _ DNSNameResolver = &ExternalEgressDNS{}
+
+// dnsResolvedName contains details about a DNS resolved name.
+type dnsResolvedName struct {
+	// dnsAddressSet gives the address set that contains the current IPs.
+	dnsAddressSet addressset.AddressSet
+	// namespaces gives those namespaces in which the DNS name is used
+	// in the EgressFirewall rules.
+	namespaces sets.Set[string]
```

Comment:
I think we need that to ensure address set is not deleted before the referencing acl is deleted, if so, it would be nice to add that to the comment

Reply:
Added the comment explaining that the address set will not be deleted before the referencing acl is deleted.